File: Src/Runtime/FileSystem/Cwd.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/27/26.
//

#include "Cwd.h"

#include <filesystem>

#include "ADT/Exception/CouldNotOpenFile.h"
#include "Support/Runtime/AccessString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::FileSystem;

ADT::Runtime::Object FileSystem::Cwd(
    ADT::List::Object &_,
    ADT::Lang::AST *__
)
{
    auto cwd = std::filesystem::current_path();;
    auto str = cwd.string();

    return {
        ADT::Runtime::ObjectType::String,
        std::string(
            str.data(),
            str.size()
        )
    };

}

File: src/runtime/file_system/cwd.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::adt::error::RuntimeError;
use crate::adt::lang::{ASTValue, RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::support::runtime::execution::continue_execution;

pub fn cwd<'a>(
    _: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let path = std::env::current_dir();
    if path.is_err() {
        return Err(
            RuntimeError::could_not_read(trace)
        );
    }
    
    let path = path.unwrap();
    let path = path.to_str();
    
    if path.is_none() {
        return Err(
            RuntimeError::could_not_read(trace)
        );   
    }
    
    let path = path.unwrap();
    continue_execution(
        Object::Hashable(
            HashableObject::String(
                ASTValue::Owned(path.to_string())
            )
        )
    )
}